core/src/mender-utils.c

```diff
@@ -509,3 +509,198 @@ mender_utils_append_list(mender_key_value_list_t **list1, mender_key_value_list_
+                *list1 = item2;
+            }
+
+            /* Update the last_item1 to the newly added node */
+            last_item1 = item2;
+            /* Move to the next item in list2 */
+            item2 = item2->next;
+            /* Ensure the last node of list1 has next set to NULL */
+            last_item1->next = NULL;
+        } else {
+            /* Move to the next item in list2 */
+            prev_item2 = item2;
+            item2      = item2->next;
+        }
+    }
+
+    return MENDER_OK;
+}
+
+mender_err_t
+mender_utils_key_value_list_delete_node(mender_key_value_list_t **list, const char *key) {
+
+    mender_key_value_list_t *to_free = NULL;
+    mender_key_value_list_t *prev    = NULL;
+    mender_key_value_list_t *item    = *list;
+    while (NULL != item) {
+        if (0 == strcmp(item->key, key)) {
+            to_free = item;
+            if (NULL == prev) {
+                *list = item->next;
+                break;
+            } else {
+                prev->next = item->next;
+            }
+            break;
+        }
+        prev = item;
+        item = item->next;
+    }
+
+    free(to_free->key);
+    free(to_free->value);
+    free(to_free);
+    return MENDER_OK;
+}
+
+mender_err_t
+mender_utils_compare_wildcard(const char *str, const char *wildcard_str, bool *match) {
+
+    assert(NULL != str);
+    assert(NULL != wildcard_str);
+    assert(NULL != match);
+
+    size_t       substrs_pos    = 0;
+    char       **substrs        = NULL;
+    char       **tmp_substrs    = NULL;
+    size_t       escaped_pos    = 0;
+    bool         found_wildcard = false;
+    mender_err_t ret            = MENDER_FAIL;
+
+    *match = true;
+
+    char *escaped = malloc(strlen(wildcard_str) + 1);
+    if (NULL == escaped) {
+        mender_log_error("Unable to allocate memory for escaped str");
+        return MENDER_FAIL;
+    }
+
+    const size_t wildcard_str_len = strlen(wildcard_str);
+
+    /* Split str into substrs on each wildcard */
+    for (size_t i = 0; i < wildcard_str_len; i++) {
+        if ('*' == wildcard_str[i]) {
+            found_wildcard       = true;
+            escaped[escaped_pos] = '\0';
+
+            /* Allocate memory for substrs
+             * temporarily assign it to tmp_substrs in case of NULL */
+            tmp_substrs = realloc(substrs, (substrs_pos + 1) * sizeof(char *));
+            if (NULL == tmp_substrs) {
+                mender_log_error("Unable to allocate memory");
+                *match = false;
+                goto END;
+            }
+            substrs = tmp_substrs;
+
+            substrs[substrs_pos] = strdup(escaped);
+            if (NULL == substrs[substrs_pos]) {
+                mender_log_error("Unable to allocate memory");
+                *match = false;
+                goto END;
+            }
+
+            substrs_pos++;
+            escaped_pos = 0;
+        } else {
+            escaped[escaped_pos++] = wildcard_str[i];
+        }
+    }
```

Comment:
AFAIK, we don't need `escaped` here at all or we don't need `substrs`. Either create a copy of `wildcard_str` and replace the wildcards with zeroes and thus split the string in-place. Or use `strchr(wildcard_str, '*')` in combination with `strndup()` to copy the substrings from `wildcard_str` into `substrs`.